src/base64.cr

```diff
@@ -83,6 +83,31 @@ module Base64
     end
   end
 
+  private def encode_with_new_lines(io : IO, &)
+    inc = 0
+    # Base64 operates in 3-byte segments, so we use a buffer size that is a
+    # multiple of 3.
+    buffer = StaticArray(UInt8, 8193).new { 0u8 }
```

Comment:
Can you elaborate on "too expensive"? How expensive is it?

Reply:
While I'm not disputing that it's faster not to zero it out (all else being equal, it _is_ doing a subset of the work), calling an operation measured in tens of nanoseconds "way too expensive" is simply incorrect.

I benchmarked the end-to-end process of encoding data from a file to another file, using a very small input like you mentioned. No matter how many times I run it, or which order they run in, the performance is indistinguishable because opening the files is 3 orders of magnitude more expensive than zeroing out 8KB of RAM — this would be even more pronounced on cloud servers that typically have slower hard drives than my MacBook Pro.

```
➜  Code crystal run --release bench_b64_encode_io.cr
nozero  20.42k ( 48.96µs) (±10.30%)  65.2kB/op   1.00× slower
  zero  20.52k ( 48.74µs) (± 8.75%)  65.2kB/op        fastest
➜  Code crystal run --release bench_b64_encode_io.cr
nozero  20.31k ( 49.23µs) (± 8.96%)  65.2kB/op   1.03× slower
  zero  20.84k ( 47.99µs) (±10.20%)  65.2kB/op        fastest
➜  Code crystal run --release bench_b64_encode_io.cr
nozero  20.59k ( 48.56µs) (± 9.61%)  65.2kB/op        fastest
  zero  20.27k ( 49.33µs) (±11.81%)  65.2kB/op   1.02× slower
```

<details><summary>Benchmark code</summary>

```crystal
require "benchmark"

input = "data.bin"
output = "data.out"
File.open input, "w" do |file|
  5.times { file << "." }
end

Benchmark.ips do |x|
  x.report("nozero") do
    File.open input do |source|
      File.open output, "w" do |destination|
        Base64.encode source.rewind, destination.rewind, zero: false
      end
    end
  end
  x.report("zero") do
    File.open input do |source|
      File.open output, "w" do |destination|
        Base64.encode source.rewind, destination.rewind, zero: true
      end
    end
  end
end

module Base64
  def encode(data : IO, io : IO, zero : Bool)
    count = 0
    encode_with_new_lines(data, zero: zero) do |byte|
      io << byte.unsafe_chr
      count += 1
    end
  end

  private def encode_with_new_lines(io : IO, zero : Bool, &)
    inc = 0
    # Base64 operates in 3-byte segments, so we use a buffer size that is a
    # multiple of 3.
    if zero
      buffer = StaticArray(UInt8, 8193).new { 0u8 }
    else
      buffer = uninitialized UInt8[8193]
    end
    bytes = buffer.to_slice

    loop do
      bytes_read = io.read(bytes)
      break if bytes_read == 0

      to_base64(bytes[0, bytes_read], CHARS_STD, pad: bytes_read < bytes.bytesize) do |byte|
        yield byte
        inc += 1
        if inc >= LINE_SIZE
          yield NL
          inc = 0
        end
      end
    end
    if inc > 0
      yield NL
    end
  end
end
```

</details>

Reply:
Let's talk about the reasons for zeroing out. Why do you think it's necessary in the first place?
We know how much data is written to the buffer and thus can scale the slice that points into it accordingly. So it shouldn't matter whether the parts that are not written to hold zeros or garbage.

Reply:
Now in this particular case, it doesn't really matter, this is a mere buffer and we only ever access `buffer[0...bytes_read]` that have been written to by `IO#read` and must never access `bytes_read..` that contains garbage.

Reply:
> Let's talk about the reasons for zeroing out. Why do you think it's necessary in the first place?

Why do you think I think it's necessary?

> We know how much data is written to the buffer and thus can scale the slice that points into it accordingly.

How do you mean? I'm using a stack-allocated buffer to avoid the heap allocation and my understanding is that its size needs to be known at compile time.

> Benchmarks are nice but you should look at the generated assembly

By all means, inspect the assembly. Static analysis is a great tool to have in your toolbox. Dismissing actual performance numbers is an odd choice, though.

Reply:
> Why do you think I think it's necessary?

Because you do it. And I understand your comments in this thread as supportive of using the zeroing constructor instead of an uninitialised `StaticArray` as suggested in https://github.com/crystal-lang/crystal/pull/14604#discussion_r1609056243

Reply:
`StaticArray(UInt8, 8193).new(0_u8)` should work as well.

Reply:
Okay. 👍 

It's usually best to use `uninitialized StaticArray(...)` when you don't need any specific initialized data like in this case.
So the original suggestion was good.
The argument about expensiveness might have some subjectivity, but I think we can agree that zeroing is completely unnecessary here. Not doing it has the potential of improving performance. However significant that may be in the greater scheme, it's the right thing to avoid unnecessary extras when possible.

Reply:
> Dismissing actual performance numbers is an odd choice, though.

Benchmarks lied to me far too much, like swap the `x.report` around and the first one is always faster :shrug: Benchmarks are useful, but don't always tell the whole story. The assembly, at least, doesn't lie, however how hard it is to decipher :joy: